### ISSUE TITLE
GridView disappears randomly when updating the observable array binding context.

### ISSUE THREAD

**Issue by davecoffin** [PeterStaev/NativeScript-Grid-View#23]
This is a really strange one. Its only happening once upgrading to the latest {N} cli (3.1.3) and the latest grid view plugin.

There's a lot going on here, but I have ruled out a lot. Here's whats happening:

When the page loads, it populates an observable array with objects representing a photo. Then the GridView displays the photos like this:
```
<PullRefresh:PullToRefresh refresh="refreshPhotos" dock="top"> 
    <gv:GridView columns="4" items="{{ photos }}" id="photoGrid" colWidth="{{gridColWidth}}" rowHeight="{{gridRowHeight}}">
    <gv:GridView.itemTemplate>
        <GridLayout>
            <GridLayout visibility="{{placeholder ? 'collapsed' : 'visible'}}" id="{{id ? id : ''}}" tap="gridViewItemTap">
                <StackLayout style.backgroundColor="#f5f5f5" />
                <Image src="{{(sizes ? sizes.sm_square : src)}}" stretch="aspectFill" propertyChange="checkImagePropertyChange" id="{{id ? id : ''}}" style="margin-bottom: 1;" visibility="{{placeholder || uploading ? 'collapsed' : 'visible'}}"></Image>
                <GridLayout visibility="{{uploading ? 'visible' : 'collapsed'}}">
                    <Image stretch="aspectFill" src="{{src ? src : src}}" style.opacity=".2" />
                    <Label text="{{percent != '100%' ? percent : 'Processing...'}}" android:marginTop="60" class="mont percent_uploaded" color="#333" height="100%" width="100%" />
                </GridLayout>
                <GridLayout visibility="{{comments.length ? 'visible' : 'collapsed'}}" class="comments_length">
                    <Label class="fa comments_length_bubble" color="#30A9FF" text="&#xf075;" />
                    <Label class="comments_length_text" color="#fff" text="{{comments.length ? comments.length : ''}}" />
                </GridLayout>
            </GridLayout>
        </GridLayout>
    </gv:GridView.itemTemplate>
</gv:GridView>  
</PullRefresh:PullToRefresh>
```

When I refresh the observable array (get fresh data from my API and update the objects in the array), it works most of the time, but at some point the gridview content just disappears. Here's a video of it happening: https://cl.ly/lpJp

You'll notice in the video that once the content disappears, it still scrolls and behaves as though the content is in there, its just not visible. Also, you may notice that I navigated to a photo, then it happened when I went back to the photos gridview. That is not a pattern, its totally random. It happens if I just continually refresh, sometimes it takes 5 refreshes, sometimes it takes 30.

I cant nail down a way to reproduce it other than just refreshing forever. Here's what is happening when the pull to refresh is called: (and by the way, i got rid of Pull to refresh and tried refreshing with a button to make sure it wasnt that plugin, and the same behavior exists):

```
function initPhotos(data) {
    console.log('beginning of initPhotos');
    rawPhotos = data;
    // remove deleted photos.
    var gottenPhotoIds = {};
    if (data.length) {
        for (var dd = 0; data.length > dd; dd++) gottenPhotoIds[data[dd].id] = true;
    }
    
    // remove photos that have been deleted.
    photos.forEach(function(item, index) {
        if (item.placeholder) {
            photos.splice(index, 1);
        } else if (!gottenPhotoIds[item.id] && !item.uploading) {
            photos.splice(index, 1);
        }
    });
    
    if (data.length) {
        for (var i = 0; data.length > i; i++) {
            if (!data[i].comments) data[i].comments = [];
            if (!displayedPhotos[data[i].id]) {
                data[i].timestamp = 'Taken at ' + helpers.convertTimezone(data[i].date_created, 'h:mma on MM/DD/YYYY');
                data[i].comment_count = (data[i].comments ? data[i].comments.length : 0);
                photos.push(observableModule.fromObject(data[i]))
                photosMap[data[i].id] = data[i];
                displayedPhotos[data[i].id] = true;
            } else {
                // update the comments if the comment count has changed
                photos.forEach(function(item, index) {
                    if (item.id == data[i].id) {
                        if (item.get('comments').length != data[i].comments.length) {
                            item.set('comments', data[i].comments);
                            photosMap[data[i].id].comments = data[i].comments;
                        }
                    }
                })
            }
        }
        photos.sort(function(a, b){
            if ((a.id-0) < (b.id-0)) {
                return 1;
            } else if ((a.id-0) > (b.id-0)) {
                return -1;
            }
        })
        photos.push({placeholder: true})
    } else {
        while (photos.length) photos.pop();
    }
    console.log('end of initPhotos');
}
```

`photos` is what I return to my UI, and is what is displayed in the grid view. So the initPhotos function takes the data from the server, then adds new photos it finds and updates photos that are already in the array.

Here's my package.json:
```
"nativescript": {
    "id": "com.cubbynotes.pocketnanny",
    "tns-android": {
      "version": "3.1.1"
    },
    "tns-ios": {
      "version": "3.1.0"
    }
  },
  "dependencies": {
    "moment": "^2.15.2",
    "nativescript-background-http": "^3.0.0",
    "nativescript-camera": "^3.0.1",
    "nativescript-geolocation": "3.0.0",
    "nativescript-grid-view": "^3.1.0",
    "nativescript-imagecropper": "0.0.7",
    "nativescript-imagepicker": "^3.0.2",
    "nativescript-iqkeyboardmanager": "^1.0.1",
    "nativescript-onesignal": "^1.0.7",
    "nativescript-orientation": "^1.6.1",
    "nativescript-permissions": "^1.2.3",
    "nativescript-photoviewer": "^1.2.0",
    "nativescript-platform-css": "^1.5.0",
    "nativescript-pulltorefresh": "^2.0.2",
    "nativescript-slides": "^2.2.11",
    "nativescript-telerik-ui": "^3.0.4",
    "nativescript-web-image-cache": "^4.2.2",
    "tns-core-modules": "^3.1.0"
  },
  "devDependencies": {
    "babel-traverse": "6.23.1",
    "babel-types": "6.23.0",
    "babylon": "6.15.0",
    "lazy": "1.0.11"
  }

```
Any ideas why this might be happening? Thanks!

**Comment by PeterStaev** [comment]
Hey @davecoffin , sadly really hard to say as you have quite a complete layout and data. Have you tried it on an actual iOS device? Also in general it is a really bad practice that you do so much modifications to the observable bound as the items of the GridView because every change causes the GridView to refresh and redraw itself. I would recommend making your changes to an unbound object and then replacing the observable that is bound.

**Comment by davecoffin** [comment]
Ill try that, thanks

**Comment by KingAndroid** [comment]
Hello @PeterStaev
How are you?
I have also met this issue.
Could you please have a look this video capture?

https://www.dropbox.com/s/4zlhz6xua1eerhw/gridview%20issue.mp4?dl=0

it is related with #14 too.
Thank you.

**Comment by PeterStaev** [comment]
Hey @KingAndroid , can you create a reproducable sample app and attach it here. I have not experienced any of the problems so not much I can help you with.

**Comment by adolfos94** [comment]
I have the same issue as @davecoffin. Have you found a fix ? I didn't try the app on a real device, but I think this only happens on iOS, I tried my application on Android and it works well. But, in iOS simulator I have this issue.

**Comment by davecoffin** [comment]
@adolfos94 I havent found a fix, I unfortunately had to switch to a different layout entirely because I couldn't make this work consistently.

**Comment by adolfos94** [comment]
@davecoffin What did you use instead of GridView ? I tried with a plain GridLayout and I couldn't make it work.

Regards.

**Comment by PeterStaev** [comment]
Hey guys, in the new 3.2 version of the widget there is a fix for #25 which might be the same problem as described here. Can you please try the new version and see if that will resolve this issue?

**Comment by PeterStaev** [comment]
No further response so closing this one for now. In case you still have problems, please provide more details.
